src/prepare_pcap.c

```diff
@@ -274,13 +279,14 @@ int prepare_pkts(const char* file, pcap_pkts* pkts)
         if (pktlen > PCAP_MAXPACKET) {
             ERROR("Packet size is too big! Recompile with bigger PCAP_MAXPACKET in prepare_pcap.h");
         }
+        /* BUG: inefficient */
         pkts->pkts = (pcap_pkt *)realloc(pkts->pkts, sizeof(*(pkts->pkts)) * (n_pkts + 1));
         if (!pkts->pkts)
             ERROR("Can't re-allocate memory for pcap pkt");
         pkt_index = pkts->pkts + n_pkts;
         pkt_index->pktlen = pktlen;
         pkt_index->ts = pkthdr->ts;
-        pkt_index->data = (unsigned char *) malloc(pktlen);
+        pkt_index->data = (unsigned char *) malloc(pktlen); /* BUG: inefficient */
```

Comment:
Yeah, I looked at refactoring this when I merged the code in initially. Problem is that you don't have enough information for the packet generation until the last minute, unlike with pcap_play - at least without some deeper work.